shell/platform/linux/fl_application.cc

```diff
@@ -0,0 +1,189 @@
+// Copyright 2024 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_application.h"
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_dart_project.h"
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_view.h"
+
+#include <gmodule.h>
+#include <gtk/gtk.h>
+
+#ifdef GDK_WINDOWING_X11
+#include <gdk/gdkx.h>
+#endif
+
+struct FlApplicationPrivate {
+  char** dart_entrypoint_arguments;
+  gboolean activated;
+};
+#define FL_APPLICATION_PRIVATE(val) static_cast<FlApplicationPrivate*>(val)
+
+enum {
+    VIEW_CONSTRUCTED,
+    LAST_SIGNAL
+};
+
+static guint signals[LAST_SIGNAL] = {0};
+
+G_DEFINE_TYPE_WITH_CODE(FlApplication,
+                        fl_application,
+                        GTK_TYPE_APPLICATION,
+                        G_ADD_PRIVATE(FlApplication))
+
+// Implements GApplication::activate.
+static void fl_application_activate(GApplication* application) {
+  FlApplication* self = FL_APPLICATION(application);
+  FlApplicationPrivate* priv =
+      FL_APPLICATION_PRIVATE(fl_application_get_instance_private(self));
+  
+  GtkApplicationWindow* app_window =
+      GTK_APPLICATION_WINDOW(gtk_application_window_new(GTK_APPLICATION(self)));
+  GtkWindow* window = GTK_WINDOW(app_window);
+
+  // Ensure we only run this function once. We currently don't do anything fancy
+  // like signaling the application that it should show itself again. 
+  // This only stops the same instance of GApplication creating multiple
+  // windows/views.
+  if (priv->activated) {
```

Comment:
I don't see why this is required - activate() is only ever called once isn't it?

Reply:
If the application is guaranteed to be unique (by not giving GApplication the flag non-unique), it will send the argv to the main instance over the dbus and activate it again in the main-instance. The gtk4 template in gnome-builder can be used to reproduce this. 

Other things can also activate it.

Reply:
Looking at all existing GTK apps I can find they don't seem to do this - if you call activate a second instance of the main window is shown, which is what Flutter applications do as well. I think we should match the existing behaviour and propose this as a second change if it's required.